crates/ra_ide/src/completion/complete_keyword.rs

```diff
@@ -121,13 +111,51 @@ fn complete_return(
 
 #[cfg(test)]
 mod tests {
-    use crate::completion::{test_utils::do_completion, CompletionItem, CompletionKind};
+    use crate::{
+        completion::{
+            test_utils::{do_completion, do_completion_with_position},
+            CompletionItem, CompletionKind,
+        },
+        CompletionItemKind,
+    };
     use insta::assert_debug_snapshot;
+    use rustc_hash::FxHashSet;
 
     fn do_keyword_completion(code: &str) -> Vec<CompletionItem> {
         do_completion(code, CompletionKind::Keyword)
     }
 
+    fn assert_completion_keyword(code: &str, keywords: &[(&str, &str)]) {
+        let (position, completion_items) =
+            do_completion_with_position(code, CompletionKind::Keyword);
+        let mut set = FxHashSet::<(String, String)>::default();
+        for (key, value) in keywords {
+            set.insert(((*key).to_string(), (*value).to_string()));
+        }
+
+        for item in completion_items {
+            assert!(item.text_edit().len() == 1);
+            assert!(item.kind() == Some(CompletionItemKind::Keyword));
+            let atom = item.text_edit().iter().next().unwrap().clone();
+            assert!(atom.delete.start() == position.offset);
+            assert!(atom.delete.end() == position.offset);
+            let pair = (item.label().to_string(), atom.insert);
+            assert!(set.contains(&pair));
+            set.remove(&pair);
+        }
+        assert!(set.is_empty());
+    }
+
+    #[test]
+    fn completes_keywords_in_use_stmt_new_approach() {
+        assert_completion_keyword(
```

Comment:
Hi @matklad! Do you think it makes sense that we use new testing approach? I will have to write a lot of tests for new keywords, so it would be good to simplify it a little bit.

Reply:
I propose to leave only labels and inserts as parameters and all other asserts do under the hood.

Reply:
Actually I need to make comparisons better so we would have a diff in the end, I will return with changes.

Reply:
One of the idea that because we have only one position to insert which we know beforehand, we can maybe generate expected completion items using this position and then just do snapshot_assert but with generated completions.

Reply:
Actually I've tried to play a little bit with asserts, but the problem is that I cannot get the assert_debug_snapshot work with arbitrary sets of completion items (and probably it _shoudn't_ work because snapshot should be somewhat static). But the thing which is good about snapshot testing is that it gives you nice diff which you would otherwise have to implement yourself.

I am a little bit worried that right now our keyword tests are somewhat not readable due to a lot of extra information. 

I see that we can do two things:
(1) Not use debug snapshots and use custom logic
(2) Use debug snapshots working not with `CompletionItems` but with simpler structures like tuples `r###"[("impl", "impl"),("struct", "struct")]"###`, that way we would still have the diffs but make the tests easier to write

Reply:
Yup, you are spot-on, this is a long-standing issue I want to fix in our completion testing infra.

Generally, there are two things we want to test during completion:

* that we complete the right set of items (this corresponds to the completion popup the user sees in the UI)
* that, each item carries the correct edit (this correponds to the user selecting a completion variant and seeing properly inserted)

We at the moment test both, and that's not really convenient. 

Ideally, the stuff in `complete_xxx` tests only the list of things, and the sutff in `presentation.rs` tests the full edits. 

I think we need to introduce two new testing utilities, and start using them for completion.

* `get_completions`, which returns a single renderd list of labels, which should look rougthly like this, but in text:

![image](https://user-images.githubusercontent.com/1711539/84510645-af6e5600-acc5-11ea-91cc-5a48678f7bbb.png)

```
v acc
M alloc
m asm!
M ast
e ABI
```

that is, just a list of labels with "icons"
* `get_single_completion(label: &str)` which additionally filters the list of completions by label, and returns a single one (which is fully feeded into insta comparision).

In this PR, I think it makes sense to do "completion list" thing.